src/mcts/search.cc

```diff
@@ -1048,285 +1151,624 @@ void SearchWorker::GatherMinibatch() {
-        // are sorted in policy decreasing order.
-        can_exit = true;
+      vtp_last_filled.push_back(-1);
+
+      // Cache all constant UCT parameters.
+      // When we're near the leaves we can copy less of the policy, since there
+      // is no way iteration will ever reach it.
+      // TODO: This is a very conservative formula. It assumes every visit we're
+      // aiming to add is going to trigger a new child, and that any visits
+      // we've already had have also done so and then a couple extra since we go
+      // to 2 unvisited to get second best in worst case.
+      // Unclear we can do better without having already walked the children.
+      // Which we are putting off until after policy is copied so we can create
+      // visited policy without having to cache it in the node (allowing the
+      // node to stay at 64 bytes).
+      int max_needed = std::min(static_cast<int>(node->GetNumEdges()),
+                                node->GetNStarted() + cur_limit + 2);
+      node->CopyPolicy(current_pol, max_needed);
+      for (int i = 0; i < max_needed; i++) {
+        current_util[i] = std::numeric_limits<float>::lowest();
+      }
+      // Root depth is 1 here, while for GetDrawScore() it's 0-based, that's why
+      // the weirdness.
+      const float draw_score = ((current_path.size() + base_depth) % 2 == 0)
+                                   ? odd_draw_score
+                                   : even_draw_score;
+      m_evaluator.SetParent(node);
+      float visited_pol = 0.0f;
+      for (Node* child : node->VisitedNodes()) {
+        int index = child->Index();
+        visited_pol += current_pol[index];
+        float q = child->GetQ(draw_score);
+        current_util[index] = q + m_evaluator.GetM(child, q);
+      }
+      const float fpu =
+          GetFpu(params_, node, is_root_node, draw_score, visited_pol);
+      for (int i = 0; i < max_needed; i++) {
+        if (current_util[i] == std::numeric_limits<float>::lowest()) {
+          current_util[i] = fpu + m_evaluator.GetDefaultM();
+        }
       }
-    }
 
-    if (second_best_edge) {
-      int estimated_visits_to_change_best =
-          best_edge.GetVisitsToReachU(second_best, puct_mult, best_without_u);
-      // Only cache for n-2 steps as the estimate created by GetVisitsToReachU
-      // has potential rounding errors and some conservative logic that can push
-      // it up to 2 away from the real value.
-      node->UpdateBestChild(best_edge,
-                            std::max(0, estimated_visits_to_change_best - 2));
-      collision_limit =
-          std::min(collision_limit, estimated_visits_to_change_best);
-      assert(collision_limit >= 1);
-      second_best_edge.Reset();
-    }
+      const float cpuct = ComputeCpuct(params_, node->GetN(), is_root_node);
+      const float puct_mult =
+          cpuct * std::sqrt(std::max(node->GetChildrenVisits(), 1u));
+      int cache_filled_idx = -1;
+      while (cur_limit > 0) {
```

Comment:
When there are more than a bunch of visits to distribute, it's likely that using a heap will be more efficient.
For most of the leaf nodes probably not though.

Here is me trying to do this in a toy example: https://github.com/mooskagh/lc0/blob/lc2v3/src/lc2/search/nodes-worker.cc#L120

Reply:
heap scales better in worst case, but the average number of loops here is just too low in practice I think.  My cpu profiles show the vast majority of CPU cost is before this loop starts - incurred bringing everything into cache from memory. Inside this loop the actual inner loop is not the most expensive part anymore - and the rest of the logic doesn't improve with a heap.
The start of the inner loop is trivially cheap except for the first time we see a new edge - but that cost can't be avoided.
So this parts of the loop we potentially avoid with heap is now just reading from some smallish stack local arrays that stay hot and in the cache and running a couple of if statements to update best/second best.  Its quite cheap.
Rather than incurring the heap complexity I had considered pre-populating 'best' with 'second best' from the previous loop.  That way the if statements fire less often.  But as I said its so cheap its not been a priority for optimization.